### PR TITLE
Simplify DoubleBucketQueue slightly

### PR DESCRIPTION
**Pull request by dnesbitt61**
Make DoubleBucketQueue and BucketQueue (meili) more common. 

In DoubleBucketQueue: Switch from deque to vector. Simplify (borrowing some logic from meili BucketQueue) by adding an empty method, using std::find to find prior label, and simply copying the tmp vector to the overflow bucket during emptyoverflow.

In BucketQueue: It seems that costmap_ could be removed if the decrease method accepts the prior cost (like it does in DoubleBucketQueue) or uses a labelcost_ functor. Also noted that the cost method is used in routing tests - not sure if this is really needed?

@ptpt - please review (and thanks for the BucketQueue code this borrows from).

### CODE REVIEW

**Comment by ptpt** [comment]
Thanks for the PR. Would be nice to hold on this PR a while; I will give it a look this weekend.

If possible I would like to migrate the tests for `BucketQueue` to work for `DoubleBucketQueue` as well https://github.com/valhalla/valhalla/blob/master/test/routing.cc

**Comment by dnesbitt61** [comment]
That sounds fine. One thing is that the DoubleBucketQueue does not hold the costs in an internal map. It relies on a functor (labelcost_) passed into the constructor - from there it is possible to access costs (which are stored in an external array indexed by the label or key). I see that BucketQueue has the ability to grow if the cost exceeds the bucket size - I thought of something similar that could also erase the front buckets as the queue grows (to keep memory in check) - but the overflow bucket seemed a better compromise.

File: src/baldr/double_bucket_queue.cc [hunk 1]
```diff
@@ -49,79 +50,63 @@ void DoubleBucketQueue::clear() {
...
   // if old cost and the new cost are in the same buckets.
-  auto& prevbucket = get_bucket(previouscost);
-  auto& newbucket  = get_bucket(newcost);
+  bucket_t& prevbucket = get_bucket(labelcost_(label));
+  bucket_t& newbucket  = get_bucket(newcost);
   if (prevbucket != newbucket) {
-    // Remove the label index from the old bucket and add to end of newbucket
+    // Add label to newbucket and remove from previous bucket
+    newbucket.push_back(label);
     for (auto it = prevbucket.begin(); it != prevbucket.end(); ++it) {
```
Comment (ptpt) [inline src/baldr/double_bucket_queue.cc]: Alternative: `std::erase(std::remove(prevbucket.begin(), prevbucket.end(), label));`

**Comment by kevinkreiser** [comment]
:rocket:

**Pull request merged**